<a href="https://colab.research.google.com/github/gsanchez21/Grethel-CAI1001C/blob/main/01-Welcome-to-AI/Guided-Project/GP01-Python-Basics-Demo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Chapter 1: Python Basics Demo
## Variables, Lists, and Basic Analysis

**AI Thinking: A Hands-On Introduction to Artificial Intelligence**

---

## What We'll Cover

This notebook demonstrates the Python concepts introduced in Chapter 1:
- Variables and data types
- Lists and basic operations
- Simple calculations (averages, totals)
- Print statements and output
- Basic loops (for loops)

You'll run each cell and see how Python works with real examples!

---

## Part 1: Variables - Storing Information

Variables are like labeled boxes that store information.

In [9]:
# Creating variables
student_name = "Grethel" # string
age = 46 # integer
gpa = 3.9 # float
is_enrolled = False # Boolean

# Display the variables
print("Student Information:")
print(f"Name: {student_name}")
print(f"Age: {age}")
print(f"GPA: {gpa}")
print(f"Enrolled: {is_enrolled}")

Student Information:
Name: Grethel
Age: 46
GPA: 3.9
Enrolled: False


### 💡 Try It Yourself

Modify the cell above to use YOUR information:
- Change the name
- Change the age
- Change the GPA

Then run it again!

In [10]:
# Variables can change!
temperature = 75
print(f"Morning temperature: {temperature}°F")

temperature = 85  # Afternoon gets hotter
print(f"Afternoon temperature: {temperature}°F")

temperature = 72  # Evening cools down
print(f"Evening temperature: {temperature}°F")

Morning temperature: 75°F
Afternoon temperature: 85°F
Evening temperature: 72°F


### 📝 What's Happening?

- `=` assigns a value to a variable
- Variables can be reassigned (changed)
- `f"..."` creates formatted strings that can include variable values using `{}`

---

## Part 2: Basic Math with Python

Python can do math like a calculator.

In [11]:
# Basic arithmetic
price = 15.99
quantity = 3

# Calculate total
total = price * quantity
print(f"Price per item: ${price}")
print(f"Quantity: {quantity}")
print(f"Total: ${total}\n")

# Add sales tax (7%)
tax_rate = 0.07
tax = total * tax_rate
final_total = total + tax

print(f"Tax (7%): ${tax:.2f}")
print(f"Final Total: ${final_total:.2f}")

Price per item: $15.99
Quantity: 3
Total: $47.97

Tax (7%): $3.36
Final Total: $51.33


### 📝 What's Happening?

- `*` multiplies numbers
- `+` adds numbers
- `.2f` in the f-string formats numbers to 2 decimal places (for money)

In [12]:
# Calculate average temperature for the week
monday = 65
tuesday = 65
wednesday = 73
thursday = 78
friday = 83

# Add them up
total_temp = monday + tuesday + wednesday + thursday + friday

# Divide by number of days
average_temp = total_temp / 5

print(f"Total temperature: {total_temp}°F")
print(f"\nAverage temperature: {average_temp}°F")

Total temperature: 364°F

Average temperature: 72.8°F


### 💡 Try It Yourself

Change the temperature values to this week's actual Miami temperatures and see the new average!

---

## Part 3: Lists - Storing Multiple Values

Lists hold multiple items in order.

In [13]:
# Create a list of temperatures
temperatures = [75, 82, 79, 85, 88]

print("This week's temperatures:")
print(temperatures)

# Access individual items (Python counts from 0!)
print(f"\nFirst day: {temperatures[0]}°F")
print(f"Second day: {temperatures[1]}°F")
print(f"Last day: {temperatures[4]}°F")

This week's temperatures:
[75, 82, 79, 85, 88]

First day: 75°F
Second day: 82°F
Last day: 88°F


### 📝 What's Happening?

- Square brackets `[]` create a list
- Lists can hold multiple values
- Access items with `list_name[index]`
- **Python counts from 0**, so the first item is index 0!

In [14]:
# List of student names
students = ["Carlos", "Maria", "Sofia", "Roberto", "Ana"]

print("Students in class:")
print(students)

print(f"\nTotal students: {len(students)}")
print(f"First student: {students[0]}")
print(f"Last student: {students[-1]}")  # -1 means last item!

Students in class:
['Carlos', 'Maria', 'Sofia', 'Roberto', 'Ana']

Total students: 5
First student: Carlos
Last student: Ana


### 💡 Cool Trick

- `len()` tells you how many items in a list
- Negative indices count backwards: `[-1]` is the last item!

---

## Part 4: Analyzing Lists

Python has built-in functions to analyze lists of numbers.

In [15]:
# Quiz scores
scores = [85, 92, 78, 95, 88]

print("Quiz Scores:")
print(scores)

# Built-in functions for analysis
total = sum(scores)          # Add all scores
average = total / len(scores) # Calculate average
highest = max(scores)         # Find highest score
lowest = min(scores)          # Find lowest score

print(f"\nTotal points: {total}")
print(f"Average score: {average:.1f}")
print(f"Highest score: {highest}")
print(f"Lowest score: {lowest}")

Quiz Scores:
[85, 92, 78, 95, 88]

Total points: 438
Average score: 87.6
Highest score: 95
Lowest score: 78


### 📝 What's Happening?

- `sum()` adds all numbers in a list
- `len()` counts items in a list
- `max()` finds the largest number
- `min()` finds the smallest number

These are the building blocks of data analysis!

In [16]:
# Monthly expenses (in dollars)
expenses = [2143, 800, 525, 510, 200, 200]

# Calculate monthly budget info
total_expenses = sum(expenses)
average_expense = total_expenses / len(expenses)
highest_expense = max(expenses)

print("Monthly Expense Analysis:")
print(f"Total expenses: ${total_expenses:.2f}")
print(f"Average per category: ${average_expense:.2f}")
print(f"Highest expense: ${highest_expense:.2f}")

Monthly Expense Analysis:
Total expenses: $4378.00
Average per category: $729.67
Highest expense: $2143.00


### 💡 Try It Yourself

Change the expenses list to represent:
- Rent
- Food
- Transportation
- Utilities
- Entertainment
- Other

Use your own estimated monthly costs!

---

## Part 5: For Loops - Processing Lists

Loops let you do something with each item in a list.

In [17]:
# List of daily temperatures
temperatures = [75, 82, 79, 85, 88]
days = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday"]

print("This Week's Weather:")
print()

# Loop through both lists at the same time
for list in range(len(days)):
    day = days[list]
    temp = temperatures[list]
    print(f"{day}: {temp}°F")

This Week's Weather:

Monday: 75°F
Tuesday: 82°F
Wednesday: 79°F
Thursday: 85°F
Friday: 88°F


### 📝 What's Happening?

- `for i in range(len(days)):` creates a loop that runs 5 times (0, 1, 2, 3, 4)
- Each time through, `i` changes (0, then 1, then 2, etc.)
- We use `i` to get the item from each list
- The indented code runs each time through the loop

In [18]:
# Simpler way - loop directly through a list
fruits = ["mango", "papaya", "guava", "maracuyá"]

print("Fruits I love:")
for fruit in fruits:
    print(f"  - {fruit}")

Fruits I love:
  - mango
  - papaya
  - guava
  - maracuyá


### 📝 What's Happening?

- `for fruit in fruits:` loops through each item
- Each time through, `fruit` gets the next item
- First time: `fruit = "mango"`
- Second time: `fruit = "papaya"`
- And so on...

In [19]:
# Calculate and print individual percentages
scores = [85, 92, 78, 95, 88]
total_possible = 100

print("Quiz Percentages:")
for score in scores:
    percentage = (score / total_possible) * 100

    # Determine the label based on the score
    if percentage >= 90:
        label = "Excellent"
    elif percentage >= 80:
        label = "Good"
    elif percentage >= 70:
        label = "Passing"
    else:
        label = "Failing"

    print(f"Score {score} = {percentage:.1f}% - {label}")

Quiz Percentages:
Score 85 = 85.0% - Good
Score 92 = 92.0% - Excellent
Score 78 = 78.0% - Passing
Score 95 = 95.0% - Excellent
Score 88 = 88.0% - Good


### 💡 Try It Yourself

Modify the loop to also print if the score is:
- "Excellent" if ≥ 90
- "Good" if ≥ 80
- "Passing" if ≥ 70

---

## Part 6: Combining Everything - A Complete Example

Let's use variables, lists, loops, and calculations together!

In [20]:
# Miami Beach Weekly Temperature Analysis

# Data for one week
days = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
temperatures = [75, 82, 79, 85, 88, 91, 87]

print("="*50)
print("MIAMI BEACH WEEKLY TEMPERATURE REPORT")
print("="*50)

# Display each day
print("\nDaily Temperatures:")
for i in range(len(days)):
    print(f"  {days[i]}: {temperatures[i]}°F")

# Calculate statistics
total_temp = sum(temperatures)
average_temp = total_temp / len(temperatures)
highest_temp = max(temperatures)
lowest_temp = min(temperatures)

print("\n" + "-"*50)
print("Weekly Statistics:")
print("-"*50)
print(f"Average temperature: {average_temp:.1f}°F")
print(f"Highest temperature: {highest_temp}°F")
print(f"Lowest temperature: {lowest_temp}°F")
print(f"Temperature range: {highest_temp - lowest_temp}°F")

# Count hot days (85°F or higher)
hot_days = 0
for temp in temperatures:
    if temp >= 85:
        hot_days += 1

print(f"\nHot days (85°F+): {hot_days} out of {len(days)}")
print("="*50)

MIAMI BEACH WEEKLY TEMPERATURE REPORT

Daily Temperatures:
  Monday: 75°F
  Tuesday: 82°F
  Wednesday: 79°F
  Thursday: 85°F
  Friday: 88°F
  Saturday: 91°F
  Sunday: 87°F

--------------------------------------------------
Weekly Statistics:
--------------------------------------------------
Average temperature: 83.9°F
Highest temperature: 91°F
Lowest temperature: 75°F
Temperature range: 16°F

Hot days (85°F+): 4 out of 7


### 📝 What's Happening?

This example combines:
1. **Two lists** - days and temperatures
2. **A loop** - to display each day
3. **Calculations** - sum, average, max, min
4. **Another loop with a counter** - counting hot days
5. **Formatted output** - making it look professional

This is the kind of data analysis you'll do with AI!

---

## Part 7: Practice Exercise

Now it's your turn! Analyze your own data.

In [22]:
# Your turn: Analyze your weekly spending

# Replace these with your actual weekly spending
spending = [40.25, 35, 20.75, 50.00, 25.25, 15.00, 45.00]
days = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]

print("MY WEEKLY SPENDING ANALYSIS")
print("="*50)

# TODO: Display each day's spending
print("\nDaily Spending:")
for i in range(len(days)):
    print(f"  {days[i]}: ${spending[i]:.2f}")

# TODO: Calculate total spending
total_spending = sum(spending)
print(f"\nTotal spent this week: ${total_spending:.2f}")

# TODO: Calculate average daily spending
average_spending = total_spending / len(spending)
print(f"Average per day: ${average_spending:.2f}")

# TODO: Find highest and lowest spending days
highest = max(spending)
lowest = min(spending)
print(f"\nMost spent in one day: ${highest:.2f}")
print(f"Least spent in one day: ${lowest:.2f}")

# BONUS: Project monthly spending
monthly_projection = total_spending * 4  # 4 weeks in a month
print(f"\nProjected monthly spending: ${monthly_projection:.2f}")

MY WEEKLY SPENDING ANALYSIS

Daily Spending:
  Monday: $40.25
  Tuesday: $35.00
  Wednesday: $20.75
  Thursday: $50.00
  Friday: $25.25
  Saturday: $15.00
  Sunday: $45.00

Total spent this week: $231.25
Average per day: $33.04

Most spent in one day: $50.00
Least spent in one day: $15.00

Projected monthly spending: $925.00


### 💡 Extension Ideas

Want to practice more? Try adding:

1. **Count expensive days** - How many days did you spend more than $40?
2. **Weekend vs. Weekday** - Compare average spending
3. **Budget check** - If your weekly budget is $200, are you over or under?
4. **Categories** - Create separate lists for food, transport, entertainment

---

## Summary: What You Learned

In this demo, you practiced:

✅ **Variables** - Storing single values with descriptive names

✅ **Data Types** - Numbers (integers and floats), text (strings), booleans

✅ **Lists** - Storing multiple related values

✅ **Indexing** - Accessing items in a list with `[0]`, `[1]`, `[-1]`

✅ **Math** - Using `+`, `-`, `*`, `/` for calculations

✅ **Built-in Functions** - `sum()`, `len()`, `max()`, `min()`

✅ **For Loops** - Processing each item in a list

✅ **Print Statements** - Displaying output with f-strings

✅ **Formatting** - Using `.2f` for decimals, creating professional output

---

## What's Next?

These are the **fundamental building blocks** of programming!

In upcoming chapters, you'll learn:
- **If statements** - Making decisions in code
- **Dictionaries** - More powerful data organization
- **Functions** - Creating reusable code blocks
- **Pandas** - Working with real datasets
- **Machine Learning** - Using these skills to build AI models

Keep practicing! Try modifying the examples above with your own data.

---

**Great work! You're ready to move forward! 🎉**

# Task
Extend the weekly spending analysis by performing the following tasks: count the number of days your spending exceeded $40; separate your `spending` data into weekday and weekend categories, then calculate and compare the average spending for each period; compare your `total_spending` with a weekly budget of $200 and state whether you are over or under budget, including the difference; create new, hypothetical lists for `food_spending`, `transport_spending`, and `entertainment_spending` (as the current `spending` list lacks category information), then calculate and display the total for each of these new categories; finally, provide a summary of all the insights gained from this extended analysis, covering expensive days, weekday/weekend comparison, budget status, and categorized spending totals.

## Count Expensive Days

### Subtask:
Iterate through your `spending` list to count and display the number of days where your spending exceeded $40.


**Reasoning**:
To fulfill the subtask, I will write Python code that initializes a counter, iterates through the 'spending' list, increments the counter for each day where spending exceeds $40, and then prints the final count.



In [23]:
expensive_days_count = 0

print("\n--- Analyzing Expensive Days ---")
for day_spending in spending:
    if day_spending > 40:
        expensive_days_count += 1

print(f"Number of days with spending over $40: {expensive_days_count}")


--- Analyzing Expensive Days ---
Number of days with spending over $40: 3


## Analyze Weekday vs. Weekend Spending

### Subtask:
Separate your `spending` data into weekday and weekend categories, then calculate and display the average spending for each period to compare them.


**Reasoning**:
To fulfill the subtask, I will write Python code that separates the `spending` data into `weekday_spending` and `weekend_spending` lists based on the `days` list, calculates the total and average spending for each period, and then prints these averages.



In [24]:
weekday_spending = []
weekend_spending = []

print("\n--- Analyzing Weekday vs. Weekend Spending ---")
for i in range(len(days)):
    if days[i] in ["Saturday", "Sunday"]:
        weekend_spending.append(spending[i])
    else:
        weekday_spending.append(spending[i])

# Calculate totals
total_weekday_spending = sum(weekday_spending)
total_weekend_spending = sum(weekend_spending)

# Calculate averages
# Avoid division by zero if there are no weekdays/weekends in the data
average_weekday_spending = total_weekday_spending / len(weekday_spending) if len(weekday_spending) > 0 else 0
average_weekend_spending = total_weekend_spending / len(weekend_spending) if len(weekend_spending) > 0 else 0

print(f"Average weekday spending: ${average_weekday_spending:.2f}")
print(f"Average weekend spending: ${average_weekend_spending:.2f}")


--- Analyzing Weekday vs. Weekend Spending ---
Average weekday spending: $34.25
Average weekend spending: $30.00


## Perform Budget Check

### Subtask:
Compare your `total_spending` with a weekly budget of $200 and print whether you are over or under the budget, along with the difference.


**Reasoning**:
To fulfill the subtask, I will define a weekly budget, calculate the difference with the total spending, and then use an if-else statement to compare the total spending with the budget and print the appropriate message.



In [25]:
weekly_budget = 200
budget_difference = total_spending - weekly_budget

print("\n--- Budget Check ---")
if total_spending > weekly_budget:
    print(f"You are over budget by ${budget_difference:.2f}")
else:
    print(f"You are under budget by ${abs(budget_difference):.2f}")


--- Budget Check ---
You are over budget by $31.25


## Categorize Spending

### Subtask:
Create new, hypothetical lists for `food_spending`, `transport_spending`, and `entertainment_spending`, then calculate and display the total for each of these new categories.

**Reasoning**:
To fulfill the subtask of categorizing spending, I will create hypothetical lists for food, transport, and entertainment spending, calculate the sum for each, and then print these totals.



In [26]:
food_spending = [75.50, 40.00, 30.25]
transport_spending = [25.00, 15.00, 10.50]
entertainment_spending = [50.00, 20.00, 12.00]

print("\n--- Categorized Spending ---")
total_food_spending = sum(food_spending)
total_transport_spending = sum(transport_spending)
total_entertainment_spending = sum(entertainment_spending)

print(f"Total Food Spending: ${total_food_spending:.2f}")
print(f"Total Transport Spending: ${total_transport_spending:.2f}")
print(f"Total Entertainment Spending: ${total_entertainment_spending:.2f}")


--- Categorized Spending ---
Total Food Spending: $145.75
Total Transport Spending: $50.50
Total Entertainment Spending: $82.00


## Summary of Extended Analysis

### Subtask:
Provide a summary of all the insights gained from this extended analysis, covering expensive days, weekday/weekend comparison, budget status, and categorized spending totals.

### Instructions:
1. Consolidate the findings from all previous analysis steps.
2. Briefly state the number of expensive days.
3. Summarize the average spending for weekdays versus weekends.
4. Clearly state whether the weekly budget was met, and by how much.
5. Present the total spending for each of the hypothetical categories (food, transport, entertainment).
6. Conclude with a brief overall insight or recommendation based on the combined analysis.

**Reasoning**:
To fulfill the subtask of summarizing the analysis, I will print the insights gained from all previous steps, including expensive days, weekday/weekend spending, budget status, and categorized spending.



In [27]:
print("\n--- Summary of Weekly Spending Analysis ---")
print("==========================================")

print(f"\n1. Expensive Days:")
print(f"   You had {expensive_days_count} day(s) where spending exceeded $40.")

print(f"\n2. Weekday vs. Weekend Spending:")
print(f"   Average weekday spending: ${average_weekday_spending:.2f}")
print(f"   Average weekend spending: ${average_weekend_spending:.2f}")

print(f"\n3. Budget Status (Weekly Budget: ${weekly_budget:.2f}):")
if total_spending > weekly_budget:
    print(f"   You are OVER budget by ${budget_difference:.2f}.")
else:
    print(f"   You are UNDER budget by ${abs(budget_difference):.2f}.")

print(f"\n4. Categorized Spending (Hypothetical):")
print(f"   - Food: ${total_food_spending:.2f}")
print(f"   - Transport: ${total_transport_spending:.2f}")
print(f"   - Entertainment: ${total_entertainment_spending:.2f}")

print("\nOverall Insight:")
print("Based on this analysis, you had a few days with higher spending, and your overall weekly spending exceeded your budget. Focusing on reducing spending on specific categories or on expensive days could help manage your finances better.")

print("==========================================")


--- Summary of Weekly Spending Analysis ---

1. Expensive Days:
   You had 3 day(s) where spending exceeded $40.

2. Weekday vs. Weekend Spending:
   Average weekday spending: $34.25
   Average weekend spending: $30.00

3. Budget Status (Weekly Budget: $200.00):
   You are OVER budget by $31.25.

4. Categorized Spending (Hypothetical):
   - Food: $145.75
   - Transport: $50.50
   - Entertainment: $82.00

Overall Insight:
Based on this analysis, you had a few days with higher spending, and your overall weekly spending exceeded your budget. Focusing on reducing spending on specific categories or on expensive days could help manage your finances better.


## Create and Analyze Categorized Spending

### Subtask:
Create new, separate hypothetical lists for food_spending, transport_spending, and entertainment_spending. Calculate and display the total for each category.


## Final Task

### Subtask:
Provide a summary of all the new insights gained from this extended spending analysis, including expensive days, weekday/weekend comparison, budget status, and categorized spending totals.


## Summary:

### Q&A
The extended spending analysis reveals that there were 3 days where spending exceeded \$40. Average weekday spending was \$34.25, which is slightly higher than the average weekend spending of \$30.00. The total weekly spending surpassed the \$200 budget by \$31.25, indicating an over-budget status. For the hypothetical categorized spending, food spending totaled \$145.75, transport spending \$50.50, and entertainment spending \$82.00.

### Data Analysis Key Findings
*   **Expensive Days**: There were 3 days where spending exceeded \$40.
*   **Weekday vs. Weekend Spending**:
    *   Average weekday spending was \$34.25.
    *   Average weekend spending was \$30.00.
*   **Budget Status**: Total weekly spending exceeded the \$200 budget by \$31.25.
*   **Categorized Spending (Hypothetical)**:
    *   Total food spending: \$145.75.
    *   Total transport spending: \$50.50.
    *   Total entertainment spending: \$82.00.

### Insights or Next Steps
*   To manage finances better, consider strategies to reduce spending on weekdays, as the average weekday spending is higher than weekend spending.
*   Focus on identifying and potentially reducing expenses in specific categories, such as food or entertainment, or on days when spending tends to exceed \$40, to help stay within budget.
